cmd/pitr/recoverer/recoverer.go

```diff
@@ -0,0 +1,268 @@
+	ListObjects(prefix string) []string
+}
+
+type RecoverType string
+
+func New(c Config) (*Recoverer, error) {
+	s3, err := storage.NewS3(c.S3Endpoint, c.S3AccessKeyID, c.S3AccessKey, c.S3BucketName, c.S3Region, true)
+	if err != nil {
+		return nil, errors.Wrap(err, "new storage manager")
+	}
+	return &Recoverer{
+		storage:        s3,
+		recoverTime:    c.RecoverTime,
+		pxcUser:        c.PXCUser,
+		pxcPass:        c.PXCPass,
+		pxcServiceName: c.PXCServiceName,
+		recoverType:    RecoverType(c.RecoverType),
+		backupNAme:     c.BackupName,
+		localDest:      "/tmp/",
+	}, nil
+}
+
+const (
+	Latest      RecoverType = "latest"      // recover to the latest existing binlog
+	Date        RecoverType = "date"        // recover to exact date
+	Transaction RecoverType = "transaction" // recover to needed trunsaction
+	Skip        RecoverType = "skip"        // skip transactions
+)
+
+func (r *Recoverer) getHost() (string, error) {
+	cmd := exec.Command("peer-list", "-on-start=/usr/bin/get-pxc-state", "-service="+r.pxcServiceName)
+	out, err := cmd.CombinedOutput()
+	if err != nil {
+		return "", errors.Wrap(err, "get output")
+	}
+	nodes := strings.Split(string(out), "node:")
+	sort.Strings(nodes)
+	lastHost := ""
+	for _, node := range nodes {
+		if strings.Contains(node, "wsrep_ready:ON:wsrep_connected:ON:wsrep_local_state_comment:Synced:wsrep_cluster_status:Primary") {
+			nodeArr := strings.Split(node, ".")
+			lastHost = nodeArr[0]
+		}
+	}
+	if len(lastHost) == 0 {
+		return "", errors.New("cant find host")
+	}
+
+	return lastHost + "." + r.pxcServiceName, nil
+}
+
+func (r *Recoverer) Run() error {
+	startGTID, err := r.GetLastBackupGTID()
+	if err != nil {
+		return errors.Wrap(err, "get last gtid")
+	}
+	err = r.SetBinlogs(startGTID)
+	if err != nil {
+		return errors.Wrap(err, "get binlog list")
+	}
+	err = r.Recover()
+	if err != nil {
+		return errors.Wrap(err, "recover")
+	}
+	return nil
+}
+
+func (r *Recoverer) DownloadBinlogs() error {
+	for _, binlog := range r.storage.ListObjects("") {
+		if strings.Contains(binlog, "set") {
+			continue
+		}
+		data, err := r.storage.GetObject(binlog)
+		if err != nil {
+			return errors.Wrap(err, "get object")
+		}
+		err = ioutil.WriteFile(binlog, data, 0644)
+		if err != nil {
+			return errors.Wrap(err, "write file")
+		}
+		r.binlogs = append(r.binlogs, binlog)
+	}
+	return nil
+}
+
+func (r *Recoverer) Recover() error {
+	host, err := r.getHost()
+	if err != nil {
+		return errors.Wrap(err, "get host")
+	}
+	cfgCmd := exec.Command("sh", "-c", `mc -C `+r.localDest+`mc config host add dest "${ENDPOINT:-https://s3.amazonaws.com}" "$ACCESS_KEY_ID" "$SECRET_ACCESS_KEY"`)
+	err = cfgCmd.Run()
+	if err != nil {
+		return errors.Wrap(err, "update mc config")
+	}
+	flags := ""
+
+	// TODO: add logic for all types
+	switch r.recoverType {
+	case Skip:
+		flags = "--exclude-gtids=" + r.excludingGTIDSet
+	case Transaction:
+	case Date:
+	case Latest:
+	}
+
+	for _, binlog := range r.binlogs {
```

Comment:
r.binlogs set in two methods - *Recoverer.DownloadBinlogs() & *Recoverer.Run(). But I don't see any call of any of these methods anywhere in code